kolibri/core/logger/api.py

```diff
@@ -913,4 +974,50 @@ class AttemptLogViewSet(ReadOnlyValuesViewset):
         "error",
         "masterylog",
         "sessionlog",
+        "diff__correct",
     )
+    field_map = {"diff": _attempts_diff}
+
+    def annotate_queryset(self, queryset):
+        if "diff__correct" not in queryset.query.annotations:
+            queryset = queryset.annotate(
+                diff__correct=Value(None, output_field=IntegerField())
+            )
+        return queryset
+
+    @action(detail=False)
+    def diff(self, request):
+        # protect against expensive queries from loose unsupported filtering
+        min_required_filter_sets = (
+            {"masterylog"},
+            {"content", "item"},
+            {"content", "user"},
+        )
+        for filter_set in min_required_filter_sets:
+            if all(k in request.GET for k in filter_set):
+                break
+        else:
+            return Response("Minimum filter condition not met", status=412)
+
+        # apply filters, narrow down to most recent try
+        target_attempt_logs = self.get_queryset()
+        # masterylog filter would filter it to a specific try
+        if "masterylog" not in request.GET:
+            # without masterylog filter, narrow down to most recent attempt for user+item
+            target_attempt_logs = AttemptLog.objects.filter(
+                id__in=Subquery(
+                    target_attempt_logs.order_by("-completion_timestamp")
```

Comment:
I think it would be safer to order by `end_timestamp` here - as we have no real guarantee that `completion_timestamp` is set for quizzes.

The other thing that might be good is to add test coverage for the case of multiple attemptlogs per user - I think my similar code here https://github.com/learningequality/kolibri/pull/8634/files# failed because of the difference in sort order in postgres.

Or... as I have just discovered, we can use this feature to be explicit about where we want nulls to be, TIL! https://docs.djangoproject.com/en/dev/ref/models/expressions/#using-f-to-sort-null-values

Might be good to use this in the order_by so that we know where we're putting the nulls (although should be less of an issue if using `end_timestamp` as it should never be null in practice).

Reply:
It's already testing with multiple attemptlogs per user, just nothing with null timestamps.